include/SFML/System/UniqueResource.hpp

```diff
@@ -0,0 +1,160 @@
+    /// \brief Deleted copy constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource(const UniqueResource&) = delete;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Deleted copy assignment
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource& operator=(const UniqueResource&) = delete;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Move constructor
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource(UniqueResource&& other) noexcept;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Move assignment
+    ///
+    ////////////////////////////////////////////////////////////
+    UniqueResource& operator=(UniqueResource&& other) noexcept;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Release ownership of the managed resource
+    ///
+    ////////////////////////////////////////////////////////////
+    void release();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Dispose of the resource by calling the deleter
+    ///
+    ////////////////////////////////////////////////////////////
+    void reset();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Dispose of the resource by calling the deleter
+    /// then assign to the resource
+    ///
+    ////////////////////////////////////////////////////////////
+    void reset(T&& resource);
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Access the underlying resource
+    ///
+    ////////////////////////////////////////////////////////////
+    [[nodiscard]] const T& get() const;
+
+private:
+    T       m_resource; //!< Handle to the managed resource
+    Deleter m_deleter;  //!< Functor to clean up resource
+    bool    m_delete;   //!< Whether or not a resource is currently owned
+};
+
+#include <SFML/System/UniqueResource.inl>
+
+} // namespace sf
+
+
+#endif // SFML_UNIQUERESOURCE_HPP
+
+
+////////////////////////////////////////////////////////////
+/// \class sf::UniqueResource
+/// \ingroup system
+///
+/// This class provides a wrapper around a handle to a resource
+/// with a custom deleter that cleans up that resource at the
+/// end of its lifetime.
+///
+/// Usage example:
+/// \code
+/// struct Deleter
+/// {
+///     void operator()(int)
+///     {
+///         std::cout << "Cleanup\n";
+///     }
+/// };
+///
+/// sf::UniqueResource<int, Deleter> uniqueResource;
+/// useResource(uniqueResource.get());
+/// ...
+/// // When destructing, prints "Cleanup"
+/// \endcode
+///
+/// This type is not copyeable but is moveable. If you move
+/// from this type, the deleter will never be called. This
+/// ensures that the resource it owns will never be deleted
+/// twice.
```

Comment:
> If you move from this type, the deleter will never be called.

This is confusing -- the deleter _will_ be called, just on the new and not on the moved-from instance.

Reply:
Right, that's confusing. I'll fix that.